crates/bls-crypto/src/hash_to_curve/try_and_increment.rs

```diff
@@ -74,98 +81,130 @@ where
+
+            // concatenate the message with the counter
+            let msg = &[&counter, extra_data, &message].concat();
+
+            // produce a hash with sufficient length
+            let mut candidate_hash = self.hasher.hash(domain, msg, hash_bytes)?;
+
+            // get the greatest flag by comparing the last bit with the greatest mask
+            let greatest = (candidate_hash[num_bytes - 1] & GREATEST_MASK) == GREATEST_MASK;
+
+            for i in 0..extension_degree {
+                // apply the mask to the last byte of each chunk
+                candidate_hash[(i + 1) * BASE_SIZE - 1] &= LAST_BYTE_MASK;
+            }
+
+            let possible_x = P::BaseField::deserialize(&mut &candidate_hash[..num_bytes])?;
+            if possible_x == P::BaseField::zero() {
+                continue;
+            }
+
+            if let Some(x) = get_point_from_x::<P>(possible_x, greatest) {
+                trace!(
+                    "succeeded hashing \"{}\" to curve in {} tries",
+                    hex::encode(message),
+                    c
+                );
+                end_timer!(hash_loop_time);
+
+                let scaled = x.scale_by_cofactor();
+                if scaled.is_zero() {
                     continue;
                 }
 
-                (possible_x, greatest)
-            };
-            match get_point_from_x_g1::<P>(possible_x, greatest) {
-                None => continue,
-                Some(x) => {
-                    trace!(
-                        "succeeded hashing \"{}\" to G1 in {} tries",
-                        hex::encode(message),
-                        c
-                    );
-                    end_timer!(hash_loop_time);
-                    let scaled = scale_by_cofactor_g1::<P>(&x.into_projective());
-                    if scaled.is_zero() {
-                        continue;
-                    }
-                    return Ok((scaled, c));
-                }
+                return Ok((scaled, c as usize));
             }
         }
         Err(BLSError::HashToCurveError)
     }
 }
 
-pub fn get_point_from_x_g1<P: Bls12Parameters>(
-    x: <P::G1Parameters as ModelParameters>::BaseField,
+/// computes y = sqrt(x^3+ax+b) and returns the corresponding group element
+pub fn get_point_from_x<P: SWModelParameters>(
+    x: P::BaseField,
     greatest: bool,
-) -> Option<G1Affine<P>> {
+) -> Option<GroupAffine<P>> {
     // Compute x^3 + ax + b
-    let x3b = <P::G1Parameters as SWModelParameters>::add_b(
-        &((x.square() * &x) + &<P::G1Parameters as SWModelParameters>::mul_by_a(&x)),
-    );
+    let x3b = P::add_b(&((x.square() * &x) + &P::mul_by_a(&x)));
 
     x3b.sqrt().map(|y| {
         let negy = -y;
 
         let y = if (y < negy) ^ greatest { y } else { negy };
-        G1Affine::<P>::new(x, y, false)
+        GroupAffine::<P>::new(x, y, false)
     })
 }
 
+/// Rounds `n` to the next power of 2
+fn next_power_of_two(mut n: usize) -> usize {
```

Comment:
Can we document or assert about the limitations of the method?

Reply:
Changed this with a function called `hash_length` which implements the round to nearest multiple of 256 as discussed (+ added test)